pkg/serviceaccounts/oauthclient/oauthclientregistry.go

```diff
@@ -10,27 +12,150 @@ import (
 	"k8s.io/kubernetes/pkg/serviceaccount"
 
 	scopeauthorizer "github.com/openshift/origin/pkg/authorization/authorizer/scope"
+	osclient "github.com/openshift/origin/pkg/client"
 	oauthapi "github.com/openshift/origin/pkg/oauth/api"
 	"github.com/openshift/origin/pkg/oauth/registry/oauthclient"
+	routeapi "github.com/openshift/origin/pkg/route/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/util/sets"
 )
 
 const (
 	OAuthRedirectURISecretAnnotationPrefix = "serviceaccounts.openshift.io/oauth-redirecturi."
 	OAuthWantChallengesAnnotationPrefix    = "serviceaccounts.openshift.io/oauth-want-challenges"
+
+	OAuthRedirectModelAnnotationPrefix              = "serviceaccounts.openshift.io/oauth-redirectmodel."
```

Comment:
So I dislike all of the names I have come up with for this API, so I welcome suggestions.  That being said, a redirect model refers to the following fields that could be used to build redirect URI(s):
```
uri.scheme
uri.port
uri.path
uri.host
resource.kind
resource.name
resource.group
```
The `resource` components define where to get the default redirect data from (route being the only resource we support for now).  If specified, the `uri` components are used to override the default data.  As long as the resulting URI(s) have a scheme and a host, they are considered valid.